### ISSUE TITLE
tmux fails when launched from script in pipe chain and requires eval to process argument variables

### ISSUE THREAD

**Issue by PathosV** [tmux/tmux#3454]
I have a bash wrapper script (tmuxdir) that takes in a list of directories and creates a command to open all directories in panes for a new tmux session

```bash
tmuxdir dir1 dir2 dir3
```
creates
```bash
cmd="tmux new-session -c dir1 \; split-window -c dir2 \; select-layout tiled \; split-window -c dir3 \; select-layout tiled"
or
cmd0="new-session -c dir1 \; split-window -c dir2 \; select-layout tiled \; split-window -c dir3 \; select-layout tiled"
```
One issue I've had is that passing the arguments to tmux
```bash
tmux "$cmd0"
```
results in a ```no server running on /tmp/tmux-****/default``` error. I've tried several permutations on how to expand the cmd argument and the only thing I've found to work for me is using ```eval tmux "$cmd0"```, which if possible I'd rather not use but for now my workaround is fine.

Now I'm trying to also allow my wrapper script to read arguments from a pipe. Whether or not I process the incoming pipe arguments, tmux seems to recognize that it's in a pipe chain and gives this error:
```
open terminal failed: not a terminal
```
I've confirmed that the cmd string gets created identically as well which is what lead me to believe tmux just doesn't want to launch if a pipe is involved.

* tmux version:
  * tmux 3.2
* Platform:
  * Linux x86_64
* TERM:
  * outside:xterm-256color
  * inside:screen

**Comment by krader1961** [comment]
The problem with your first issue is that you are using quotes to make the commands and their arguments a single token. See the "PARSING SYNTAX" section of the tmux man page. You are expected to pass each command and associated arguments as individual tokens. For example:

```
tmux new-session -c dir1
```

Not

```
tmux "new-session -c dir1"
```

Your use of the shell's `eval` command causes the shell to parse your quoted string into tokens separated by whitespace. Which, as you noted, is dangerous unless you can guarantee the string does not contain any unexpected shell meta-characters (like semicolon). This isn't really a tmux issue. It's a generic POSIX shell scripting question that applies to practically every other command you might want to which you want to provide multiple arguments. One solution, assuming a sufficiently new enough version of bash, is to use an array of arguments. For example,

```
cmd0=(new-session -c dir1 ";" split-window -c dir2)
tmux "${cmd0[@]}"
```

As for your last issue... Tmux expects stdin to be connected to your terminal except for a few specific cases such as this example from the man page: `make 2>&1|tmux splitw -dI &`. What you should do in your wrapper script is connect tmux's stdin to your tty: `tmux </dev/tty`.

**Comment by PathosV** [comment]
I should have listed out that I did try using an expanded array for the cmd variable, but your example pointed to why it wasnt working for me. I had everything still in double quotes so I was not getting the elements I needed, problem 1 solved, thanks!

As for connectting to tty,
```
tmux </dev/tty "${cmd0[@]}"
```
and
```
tmux  "${cmd0[@]}" </dev/tty
```
both return
```
open terminal failed: can't use /dev/tty
```
when using a pipe test. I'm not doing anything with the stdin, just making it present.
```
echo a | tmuxdir dir1 dir2 dir3
```


In either case, I don't actually need to pipe anything to tmux. I'm piping to the wrapper script which will process the stdin. At that point if its possible to detach from the pipe and cleanly launch tmux then I'm all set.

**Comment by nicm** [comment]
`new-session` will attach so it requires stdin to be a tty. If you don't want to attach use `-d`; if you can't provide a tty then use `-d` and then attach separately.

**Comment by krader1961** [comment]
> open terminal failed: can't use /dev/tty

Something is wrong with your system since redirecting from/to */dev/tty* should work on every Unix system. Alternatively, you've left out an important fact; for example, you are running your script in a context that does not have a controlling tty. This is on one of my Linux systems:

```
bash> echo yes </dev/tty
yes
bash> ls -l /dev/tty
crw-rw-rw- 1 root tty 5, 0 Jan 21 21:32 /dev/tty
```

**Comment by systemmonkey42** [comment]
> Something is wrong with your system since redirecting from/to /dev/tty should work

This error is generated by tmux.  There is no problem here.

```
tmux  "new-session" </dev/tty
open terminal failed: can't use /dev/tty
```

The special file `/dev/tty` is different to every process which opens it.  In the case of `tmux`, it calls `ttyname()` on stdin, which resolves so `/dev/tty`, forks, calls `setpgrp()`  (which starts a new process group with a new controlling terminal) then atttempts to open `/dev/tty` directly and gets an error (because the new background session does not have a tty).

Without a redirection, `ttyname()` returns something like `/dev/pts/2`, so when the server has forked and attempts to attach, it works correctly.

The bottom line is, not use redirection to tmux.  It doesn't work the way you think it does.

**Comment by krader1961** [comment]
> The special file /dev/tty is different to every process which opens it.

Exactly. It is supposed to resolve to the same tty device as the current controlling terminal (e.g., */dev/pts/2*). There is something weird about how the `ttyname()` function is doing its job as you can see from this experiment (on macOS but Linux will produce similar results as you noted) despite the fact that stdin refers to the same (logical if not physical) device in both cases.:

```
> tty
/dev/ttys000
> tty </dev/tty
/dev/tty
```

I am honestly surprised. And I spent two decades as a Unix support engineer (1985 to 2007) so I can't believe I have never noticed this before now.

> The bottom line is, don't use redirection to tmux when launching a server. It doesn't work the way you think it does.

It does work the way I think (i.e., expect) it to except with regard to the `ttyname()` function. Consider this example:

```
> cat </dev/tty
abc
[Ctrl-d]
abc
```

In that example I typed "abc", [Enter], [Ctrl-d], which caused `cat` to echo the "abc" I typed into the tty. The `cat` command read from my controlling tty even though its official name was */dev/ttys000* and not */dev/tty*. It's easy to find problem reports by other individuals who also think `ttyname()` is wrong; e.g., https://linux.debian.bugs.dist.narkive.com/x30lM5y8/bug-682972-ttyname-and-bin-tty-on-dev-tty-return-dev-tty.

**Comment by PathosV** [comment]
I've cut down my code to better explain:
```bash
#tmuxdir
if [ ! -t 0 ];then #Check if stdin has data
        dir=($(cat -))
else
        dir=($(echo $@))
fi
##PIPE/STDIN NO LONGER NEEDED##
for D in "${dir[@]}";do
        ((n++))
        if [ $n -eq 1 ];then
                cmd=(new-session -c $D)
        else
                cmd+=( \; split-window -c $D \; select-layout tiled)
        fi
done
tmux  "${cmd[@]}"
```
```
> tmuxdir dir1 dir2
~~success~~
> echo dir1 dir2 | tmuxdir
open terminal failed: not a terminal

```

So while my script can exist in a pipe to grab directory arguments, tmux does not directly interact with the pipe and I don't want it to. I plan on using pipes to handle line fed arguments but for now this example shows my issue.

**Comment by krader1961** [comment]
@PathosV: It would be simpler, and clearer, to use command capture rather than reading the info from stdin:

```
tmuxdir $(echo dir1 dir2)
```

**Comment by PathosV** [comment]
@systemmonkey42 that worked! thank you so much. I'll go ahead and close the issue.
@krader1961 I wanted to go this route so i could handle new line delimited data filtered from other outputs. With @systemmonkey42's solution I'm all set.
